xapian-core/cluster/kmeans.cc

```diff
@@ -0,0 +1,146 @@
+/** @file kmeans.cc
+ *  @brief KMeans clustering API
+ */
+/* Copyright (C) 2016 Richhiey Thomas
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#include <config.h>
+
+#include "xapian/cluster.h"
+#include <debuglog.h>
+#include <xapian/error.h>
+
+#include <set>
+#include <unordered_map>
+#include <vector>
+
+using namespace Xapian;
+using namespace std;
+
+string
+KMeans::get_description() const {
+    LOGCALL(API, string, "KMeans::get_description()", NO_ARGS);
+    return "KMeans clusterer";
+}
+
+void
+KMeans::initialize_centroids(ClusterSet &cset) {
+    LOGCALL_VOID(API, "KMeans::initialize_centroids()", cset);
+    initialize_random(cset);
+}
+
+void
+KMeans::initialize_random(ClusterSet &cset) {
+    LOGCALL_VOID(API, "KMeans::initialize_random()", cset);
+    set<docid> centroid_docs;
+    int size = docs.size();
+    srand(time(NULL));
+    for (unsigned int i = 0; i < k;) {
+	int x = rand() % size;
+	if (centroid_docs.find(x) == centroid_docs.end()) {
+	    Point a = docs[x];
+	    Centroid centroid;
+	    centroid.set_to_point(a);
+	    centroids.push_back(centroid);
+	    centroid_docs.insert(x);
+	    Cluster cluster1(centroid);
+	    cset.add_cluster(cluster1);
+	    ++i;
+	}
+    }
+}
+
+void
+KMeans::initialize_points(const MSet &source, TermListGroup &tlg) {
+    LOGCALL_VOID(API, "KMeans::initialize_points()", source | tlg);
+    for (MSetIterator it = source.begin(); it != source.end(); ++it) {
+	Point p;
+	p.initialize(tlg, it.get_document());
+	docs.push_back(p);
+    }
+}
+
+bool
+KMeans::converge(vector<Centroid> &previous, vector<Centroid> &current) const {
+    LOGCALL(API, bool, "KMeans::converge()", previous | current);
+    CosineDistance cosine;
+    for (unsigned int i = 0; i < k; ++i) {
+	Centroid &prev = previous[i];
+	Centroid &curr = current[i];
+	double dist = cosine.similarity(prev, curr);
+	if (dist > 0.0001)
```

Comment:
This is something I discovered by running it with different threshold values. The clusters do not change or tend to change very minimally once the distances between centroids gets below 0.0001. I'm not yet sure whether this is the most optimal value though.